### ISSUE TITLE
Reproducible builds expressed as Provenance predicate

### ISSUE THREAD

**Issue by MarkLodato** [in-toto/attestation#3]
There are a few open questions for how to support [reproducible builds](https:reproducible-builds.org) in the Provenance predicate:

Q1. Should each rebuilder produce its own attestation, or should all rebuilders sign the same attestation?

(option A) With the current schema, `builder` is required so each rebuilder must produce a unique attestation. The consumer would then verify that all the fields are identical except `builder`.

(option B) An alternate idea would be for all rebuilders to sign the same statement. This would only work if `builder` is optional and implicit from the signing key.

Q2. Should the Provenance attestation be sufficient to reproduce a build?

I think that's a good idea, but I'm not sure we're there yet.

At a minimum, we should document that notion and perhaps have a `reproducible: true` field to indicate that the builder thinks it is reproducible.

Even then, the rebuilder needs to understand the recipe. For example, if GitHub Actions is used, the need to understand how to parse and run the workflow. It would be nice if that we had some standardized convention for recipes, so that all builds speak the same language. Too big a project for the attestation format, but something to think about as part of SLSA.

**Comment by TomHennen** [comment]
Q1: I think they should produce their own attestation.  There aren't, yet, any fields that might differ from builder to builder, but it's possible they might be introduced (imagine if measured boot winds up in the attestation).  These differences might be immaterial, but it would make it difficult for them to sign the same statement.  Additionally, having them sign the same statement introduces an ordering problem.  Who produces the first statement?  How do different builders communicate on where the statement lives so that it can be signed?  With each builder signing their own statement there doesn't have to be any such communication.  Eventually, of course, people will want some way to aggregate these results, but that could be done by another project, potentially in different ways as is best suitable to the situation.

Q2: Yes it would be nice.  We'd probably need some way to indicate standard properties of the build environment, which may not be currently expressed in the provenance (unless I'm mistaken).

**Comment by MarkLodato** [comment]
Agreed. I'll leave this issue open to document the reproducible builds in the repo.
